pkg/controller/direct/networkmanagement/connectivitytest_controller.go

```diff
@@ -0,0 +1,353 @@
+
+	req := &pb.GetConnectivityTestRequest{Name: a.id.String()}
+	actual, err := a.gcpClient.GetConnectivityTest(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting networkmanagement connectivitytest %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *connectivityTestAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating networkmanagement connectivitytest", "name", a.id)
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := a.desired.DeepCopy()
+	resource := NetworkManagementConnectivityTestSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &pb.CreateConnectivityTestRequest{
+		Parent:   a.id.Parent().String(),
+		TestId:   a.id.ID(),
+		Resource: resource,
+	}
+	op, err := a.gcpClient.CreateConnectivityTest(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating networkmanagement connectivitytest %s: %w", a.id.String(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("networkmanagement connectivitytest %s waiting creation: %w", a.id.String(), err)
+	}
+	log.V(2).Info("successfully created networkmanagement connectivitytest in gcp", "name", a.id)
+
+	status := &krm.NetworkManagementConnectivityTestStatus{}
+	status.ObservedState = NetworkManagementConnectivityTestObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *connectivityTestAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating networkmanagement connectivitytest", "name", a.id)
+
+	if err := a.normalizeReferenceFields(ctx); err != nil {
+		return err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := a.desired.DeepCopy()
+	resource := NetworkManagementConnectivityTestSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	resource.Name = a.id.String() // Set the name for the update request
+
+	paths, err := common.CompareProtoMessage(resource, a.actual, common.BasicDiff)
+	if err != nil {
+		return fmt.Errorf("calculating diff for connectivity test %s: %w", a.id, err)
+	}
+
+	topLevelFieldPaths := sets.New[string]()
+	for path, _ := range paths {
+		tokens := strings.Split(path, ".")
+		topLevelFieldPaths.Insert(tokens[0])
+	}
+	// Remove output-only fields.
+	topLevelFieldPaths.Delete("reachability_details")
```

Comment:
Is this field incorrect labeled in the proto?

Reply:
It is actually correctly labeled: http://github.com/googleapis/googleapis/blame/master/google/cloud/networkmanagement/v1/connectivity_test.proto#L93, but somehow CompareProtoMessage() returns diff in the subfield under "reachability_details".